From 3fe37888e0e5d090dba393a078781a968e05b879 Mon Sep 17 00:00:00 2001
From: Eddie James <eajames@linux.ibm.com>
Date: Mon, 10 Aug 2020 09:59:02 -0500
Subject: [PATCH 8/8] hwclock: Fix settimeofday for glibc v2.31+

The glibc implementation changed for settimeofday, resulting in "invalid
argument" error when attempting to set both timezone and time with a single
call. Fix this by calling settimeofday twice

Signed-off-by: Eddie James <eajames@linux.ibm.com>
Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
(cherry picked from commit a4faa9af79c2a0f2b501c0fd507842e4966fc23b)
(Klaus: backported to 1.31.1 by using bb_perror_msg_and_die())
Signed-off-by: Klaus Heinrich Kiwi <klaus@linux.vnet.ibm.com>
---
 util-linux/hwclock.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/util-linux/hwclock.c b/util-linux/hwclock.c
index 29f51021e..5da250e0d 100644
--- a/util-linux/hwclock.c
+++ b/util-linux/hwclock.c
@@ -122,16 +122,20 @@ static void to_sys_clock(const char **pp_rtcname, int utc)
 	struct timeval tv;
 	struct timezone tz;
 
-	tz.tz_minuteswest = timezone/60;
+	tz.tz_minuteswest = timezone / 60;
 	/* ^^^ used to also subtract 60*daylight, but it's wrong:
 	 * daylight!=0 means "this timezone has some DST
 	 * during the year", not "DST is in effect now".
 	 */
 	tz.tz_dsttime = 0;
 
+	/* glibc v2.31+ returns an error if both args are non-NULL */
+	if (settimeofday(NULL, &tz))
+		bb_simple_perror_msg_and_die("settimeofday");
+
 	tv.tv_sec = read_rtc(pp_rtcname, NULL, utc);
 	tv.tv_usec = 0;
-	if (settimeofday(&tv, &tz))
+	if (settimeofday(&tv, NULL))
 		bb_perror_msg_and_die("settimeofday");
 }
 
@@ -283,7 +287,11 @@ static void set_system_clock_timezone(int utc)
 	gettimeofday(&tv, NULL);
 	if (!utc)
 		tv.tv_sec += tz.tz_minuteswest * 60;
-	if (settimeofday(&tv, &tz))
+
+	/* glibc v2.31+ returns an error if both args are non-NULL */
+	if (settimeofday(NULL, &tz))
+		bb_perror_msg_and_die("settimeofday");
+	if (settimeofday(&tv, NULL))
 		bb_perror_msg_and_die("settimeofday");
 }
 
-- 
2.17.1

